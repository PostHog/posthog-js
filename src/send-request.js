import { _, logger } from './utils'

export const encodePostData = (data, options) => {
    if (options.blob && data.buffer) {
        return new Blob([data.buffer], { type: 'text/plain' })
    }

    if (options.sendBeacon || options.blob) {
        const body = encodePostData(data, { method: 'POST' })
        return new Blob([body], { type: 'application/x-www-form-urlencoded' })
    }

    if (options.method !== 'POST') {
        return null
    }

    let body_data
    const isUint8Array = (d) => Object.prototype.toString.call(d) === '[object Uint8Array]'
    if (Array.isArray(data) || isUint8Array(data)) {
        body_data = 'data=' + encodeURIComponent(data)
    } else {
        body_data = 'data=' + encodeURIComponent(data['data'])
    }

    if (data['compression']) {
        body_data += '&compression=' + data['compression']
    }

    return body_data
}

export const xhr = ({
    url,
    data,
    headers,
    options,
    captureMetrics,
    callback,
    retriesPerformedSoFar,
    retryQueue,
    onXHRError,
}) => {
    const req = new XMLHttpRequest()
    req.open(options.method, url, true)

    const body = encodePostData(data, options)

    captureMetrics.incr('_send_request')
    captureMetrics.incr('_send_request_inflight')

    _.each(headers, function (headerValue, headerName) {
        req.setRequestHeader(headerName, headerValue)
    })
    if (options.method === 'POST' && !options.blob) {
        req.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded')
    }

    // send the ph_optout cookie
    // withCredentials cannot be modified until after calling .open on Android and Mobile Safari
    req.withCredentials = true
    req.onreadystatechange = () => {
        if (req.readyState === 4) {
            captureMetrics.incr(`xhr-response`)
            captureMetrics.incr(`xhr-response-${req.status}`)
            captureMetrics.decr('_send_request_inflight')

            // XMLHttpRequest.DONE == 4, except in safari 4
            if (req.status === 200) {
                if (callback) {
                    let response
                    try {
                        response = JSON.parse(req.responseText)
                    } catch (e) {
                        logger.error(e)
                        return
                    }
                    callback(response)
                }
            } else {
                if (typeof onXHRError === 'function') {
                    onXHRError(req)
                }

                // don't retry certain errors
                if ([401, 403, 404, 500].indexOf(req.status) < 0) {
                    retryQueue.enqueue({
                        url,
                        data,
                        options,
                        headers,
                        retriesPerformedSoFar: (retriesPerformedSoFar || 0) + 1,
                        callback,
                    })
                }

                if (callback) {
                    if (options.verbose) {
                        callback({ status: 0, error: error })
                    } else {
                        callback(0)
                    }
                }
            }
        }
    }
    req.send(body)
}
