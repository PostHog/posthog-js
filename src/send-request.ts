import { _each, _HTTPBuildQuery, logger } from './utils'
import Config from './config'
import { PostData, XHROptions, XHRParams } from './types'

export const addParamsToURL = (
    url: string,
    urlQueryArgs: Record<string, any> | undefined,
    parameterOptions: { ip?: boolean }
): string => {
    const args = urlQueryArgs || {}
    args['ip'] = parameterOptions['ip'] ? 1 : 0
    args['_'] = new Date().getTime().toString()
    args['ver'] = Config.LIB_VERSION

    const halves = url.split('?')
    if (halves.length > 1) {
        const params = halves[1].split('&')
        for (const p of params) {
            const key = p.split('=')[0]
            if (args[key]) {
                delete args[key]
            }
        }
    }

    const argSeparator = url.indexOf('?') > -1 ? '&' : '?'
    return url + argSeparator + _HTTPBuildQuery(args)
}

export const encodePostData = (data: PostData | Uint8Array, options: Partial<XHROptions>): string | BlobPart | null => {
    if (options.blob && data.buffer) {
        return new Blob([data.buffer], { type: 'text/plain' })
    }

    if (options.sendBeacon || options.blob) {
        const body = encodePostData(data, { method: 'POST' }) as BlobPart
        return new Blob([body], { type: 'application/x-www-form-urlencoded' })
    }

    if (options.method !== 'POST') {
        return null
    }

    let body_data
    const isUint8Array = (d: unknown): d is Uint8Array => Object.prototype.toString.call(d) === '[object Uint8Array]'
    if (Array.isArray(data) || isUint8Array(data)) {
        // TODO: eh? passing an Array here?
        body_data = 'data=' + encodeURIComponent(data as any)
    } else {
        body_data = 'data=' + encodeURIComponent(data.data as string)
    }

    if ('compression' in data && data.compression) {
        body_data += '&compression=' + data.compression
    }

    return body_data
}

export const xhr = ({
    url,
    data,
    headers,
    options,
    callback,
    retriesPerformedSoFar,
    retryQueue,
    onXHRError,
    timeout = 10000,
    onRateLimited,
}: XHRParams) => {
    const req = new XMLHttpRequest()
    req.open(options.method || 'GET', url, true)

    const body = encodePostData(data, options)

    _each(headers, function (headerValue, headerName) {
        req.setRequestHeader(headerName, headerValue)
    })

    if (options.method === 'POST' && !options.blob) {
        req.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded')
    }

    req.timeout = timeout
    // send the ph_optout cookie
    // withCredentials cannot be modified until after calling .open on Android and Mobile Safari
    req.withCredentials = true
    req.onreadystatechange = () => {
        if (req.readyState === 4) {
            // XMLHttpRequest.DONE == 4, except in safari 4
            if (req.status === 200) {
                if (callback) {
                    let response
                    try {
                        response = JSON.parse(req.responseText)
                    } catch (e) {
                        logger.error(e)
                        return
                    }
                    callback(response)
                }
            } else {
                if (typeof onXHRError === 'function') {
                    onXHRError(req)
                }

                // don't retry certain errors
                if ([401, 403, 404, 500].indexOf(req.status) < 0) {
                    retryQueue.enqueue({
                        url,
                        data,
                        options,
                        headers,
                        retriesPerformedSoFar: (retriesPerformedSoFar || 0) + 1,
                        callback,
                    })
                }

                if (req.status === 429) {
                    onRateLimited?.(req)
                }

                if (callback) {
                    callback({ status: 0 })
                }
            }
        }
    }
    req.send(body)
}
