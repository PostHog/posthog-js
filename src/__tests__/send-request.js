/* eslint-disable compat/compat */

import { addParamsToURL, encodePostData, xhr } from '../send-request'
import { assert, boolean, property, uint8Array, VerbosityLevel } from 'fast-check'

jest.mock('../config', () => ({ DEBUG: false, LIB_VERSION: '1.23.45' }))

describe('xhr', () => {
    given('mockXHR', () => ({
        open: jest.fn(),
        setRequestHeader: jest.fn,
        onreadystatechange: jest.fn(),
        send: jest.fn(),
        readyState: 4,
        responseText: JSON.stringify('something here'),
        status: 502,
    }))
    given('onXHRError', jest.fn)
    given('on429Response', jest.fn)
    given('xhrOptions', () => ({}))
    given('xhrParams', () => ({
        url: 'https://any.posthog-instance.com',
        data: '',
        headers: {},
        options: given.xhrOptions,
        callback: () => {},
        retriesPerformedSoFar: null,
        retryQueue: {
            enqueue: () => {},
        },
        onXHRError: given.onXHRError,
        onRateLimited: given.on429Response,
    }))
    given('subject', () => () => {
        xhr(given.xhrParams)
        given.mockXHR.onreadystatechange()
    })

    beforeEach(() => {
        window.XMLHttpRequest = jest.fn(() => given.mockXHR)
    })

    describe('when xhr requests fail', () => {
        it('does not error if the configured onXHRError is not a function', () => {
            given('onXHRError', () => 'not a function')
            expect(() => given.subject()).not.toThrow()
        })

        it('calls the injected XHR error handler', () => {
            //cannot use an auto-mock from jest as the code checks if onXHRError is a Function
            let requestFromError
            given('onXHRError', () => (req) => (requestFromError = req))
            given.subject()
            expect(requestFromError).toHaveProperty('status', 502)
        })

        it('calls the injected 429 handler for 429 responses', () => {
            given.mockXHR.status = 429
            given.subject()
            expect(given.on429Response).toHaveBeenCalledWith(given.mockXHR)
        })

        it('does not call the injected 429 handler for non-429 responses', () => {
            given.mockXHR.status = 404
            given.subject()
            expect(given.on429Response).not.toHaveBeenCalled()
        })
    })
})

describe('adding query params to posthog API calls', () => {
    given('subject', () => () => addParamsToURL(given.posthogURL, given.urlQueryArgs, given.parameterOptions))
    given('posthogURL', () => 'https://any.posthog-instance.com')
    given('urlQueryArgs', () => ({}))
    given('parameterOptions', () => ({
        ip: true,
    }))

    it('adds library version', () => {
        expect(new URL(given.subject()).search).toContain('&ver=1.23.45')
    })

    it('adds i as 1 when IP in config', () => {
        expect(new URL(given.subject()).search).toContain('ip=1')
    })
    it('adds i as 0 when IP not in config', () => {
        given('parameterOptions', () => ({}))
        expect(new URL(given.subject()).search).toContain('ip=0')
    })
    it('adds timestamp', () => {
        expect(new URL(given.subject()).search).toMatch(/_=\d+/)
    })

    it('does not add a query parameter if it already exists in the URL', () => {
        given('posthogURL', () => 'https://test.com/')
        const whenItShouldAddParam = given.subject()
        expect(whenItShouldAddParam).toContain('ver=1.23.45')

        given('posthogURL', () => 'https://test.com/decide/?ver=2')
        const whenItShouldNotAddParam = given.subject()
        expect(whenItShouldNotAddParam).not.toContain('ver=1.23.45')
        expect(whenItShouldNotAddParam).toContain('ver=2')
    })
})

describe('using property based testing to identify edge cases in encodePostData', () => {
    it('can use many combinations of typed arrays and options to detect if the method generates undefined', () => {
        assert(
            property(uint8Array(), boolean(), boolean(), (data, blob, sendBeacon) => {
                const encodedData = encodePostData(data, { blob, sendBeacon, method: 'POST' })
                // returns blob or string - ignore when it is not a string response
                return encodedData.indexOf && encodedData.indexOf('undefined') < 0
            }),
            { numRuns: 1000, verbose: VerbosityLevel.VeryVerbose }
        )
    })

    it('does not return undefined when blob and send beacon are false and the input is an empty uint8array', () => {
        const encodedData = encodePostData(new Uint8Array([]), { method: 'POST' })
        expect(typeof encodedData).toBe('string')
        expect(encodedData).toBe('data=')
    })
})

describe('encodePostData()', () => {
    given('subject', () => encodePostData(given.data, given.options))

    given('data', () => ({ data: 'content' }))
    given('options', () => ({ method: 'POST' }))

    beforeEach(() => {
        jest.spyOn(global, 'Blob').mockImplementation((...args) => ['Blob', ...args])
    })

    it('handles objects', () => {
        expect(given.subject).toMatchSnapshot()
    })

    it('handles arrays', () => {
        given('data', () => ['foo', 'bar'])

        expect(given.subject).toMatchSnapshot()
    })

    it('handles data with compression', () => {
        given('data', () => ({ data: 'content', compression: 'lz64' }))

        expect(given.subject).toMatchSnapshot()
    })

    it('handles GET requests', () => {
        given('options', () => ({ method: 'GET' }))

        expect(given.subject).toEqual(null)
    })

    it('handles blob', () => {
        given('options', () => ({ method: 'POST', blob: true }))
        given('data', () => ({ buffer: 'buffer' }))

        expect(given.subject).toMatchSnapshot()
    })

    it('handles sendBeacon', () => {
        given('options', () => ({ method: 'POST', sendBeacon: true }))

        expect(given.subject).toMatchSnapshot()
    })

    it('handles sendBeacon when data is not a typed array and blob is also true', () => {
        given('options', () => ({ method: 'POST', sendBeacon: true, blob: true }))

        expect(given.subject).toMatchSnapshot()
    })

    it('handles sendBeacon when data is a typed array and blob is also true', () => {
        given('data', () => new Uint8Array([1, 2, 3, 4]))
        given('options', () => ({ method: 'POST', sendBeacon: true, blob: true }))

        expect(given.subject).toMatchSnapshot()
    })
})
