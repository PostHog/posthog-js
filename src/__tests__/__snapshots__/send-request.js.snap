// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`encodePostData() handles arrays 1`] = `"data=foo%2Cbar"`;

exports[`encodePostData() handles blob 1`] = `
Array [
  "Blob",
  Array [
    "buffer",
  ],
  Object {
    "type": "text/plain",
  },
]
`;

exports[`encodePostData() handles data with compression 1`] = `"data=content&compression=lz64"`;

exports[`encodePostData() handles objects 1`] = `"data=content"`;

exports[`encodePostData() handles sendBeacon 1`] = `
Array [
  "Blob",
  Array [
    "data=content",
  ],
  Object {
    "type": "application/x-www-form-urlencoded",
  },
]
`;

exports[`encodePostData() handles sendBeacon when data is a typed array and blob is also true 1`] = `
Array [
  "Blob",
  Array [
    ArrayBuffer [],
  ],
  Object {
    "type": "text/plain",
  },
]
`;

exports[`encodePostData() handles sendBeacon when data is not a typed array and blob is also true 1`] = `
Array [
  "Blob",
  Array [
    "data=content",
  ],
  Object {
    "type": "application/x-www-form-urlencoded",
  },
]
`;
